### PR TITLE
Add Crafting Input Buffer (ME)

### PR DESCRIPTION
**Pull request by harrynull**
* This PR adds a new type of hatch/bus "Crafting Input Buffer (ME)," which is essentially the combination of input bus, input hatch and dual interface. It can contain 32 different patterns, in addition to one manual slot for holding catalyst (e.g. mold), and one circuit slot. It has infinitely large internal item/fluid buffer for each pattern slot, which helps with recipe conflicts too.
* This has not been extensively tested yet, and it's my first time doing some major gtnh modding. i.e. i dont know what im doing. So feedback/testing welcome
* This PR does not include a recipe. Suggested recipe: 1 Stocking Input Bus + 1 Input Hatch (EV) + 3 Pattern Capacity Card + 1 Dual Interface. However, if it is deemed that this is too OP and might trivialize automation, then it could be gated infinity chest, or placed in other tiers.
* This is intended to replace some of the tedious setups where you interface into an infinity chest / ME chest, and then use stocking input bus. <del>One caveat is that it is not as scalable because you cannot link multiple Crafting Input Buffer together like you could with a ME subnet.</del> Now we have crafting input slave which does this. See comment for details.
* However, if fluids are involved, then it could solve the current issue where the ingredient buffer is way too small, P2P does not stack fluids, and interface-to-interface/storage bus has a very limited throughput.
* You can also use it to solve the annoying issue of mixer recipe conflicts.
* You can finally use one LPF / PA for solidfying everything!
* It instantly does one recipe check when AE pushes items, so no more waiting for multis to start. They start crafting instantly like singleblocks.
* Also because each inventory is small, the recipe checks should theoretically be faster too. It is also possible to implement a recipe hint (basically locked recipe per pattern slot), though it is not implemented yet.
* It currently does not show up in interface terminal, and does not support blocking mode (you have to rely on AE to do the load balancing)
* It can be placed anywhere an input bus is accepted.
* It is currently only usable for multis migrated to `generic processing logic`. Which is most gt++, good generators and bartworks multi.
* The overlay texture is edited from [AE2](https://github.com/GTNewHorizons/Applied-Energistics-2-Unofficial/blob/master/src/main/resources/assets/appliedenergistics2/textures/items/ItemMaterial.BlankPattern.png) licensed under (c) 2013 - 2015 AlgorithmX2 et al CC BY-NC-SA 3.0 License.

image
(This is the original GUI design, demonstrating crafting with items only, items + circuits, item + fluid)

### CODE REVIEW

**Comment by harrynull** [comment]
Should I make it so the internal inventory of the pattern is shared across all Crafting Input Buffer in the same AE network, as long as the pattern is identical? This should solve the load balancing issue.

**Comment by FoKuCNiK** [comment]
say this is what is missing

**Comment by repo-alt** [comment]
I don't see where crafting ingredients areextracted from AE

**Comment by harrynull** [comment]
> I don't see where crafting ingredients are extracted from AE

`acceptPattern` will take everything into its internal buffer

**Comment by chochem** [comment]
for a recipe this needs to be much more late game. it trivializes a lot.

Its an even more powerful version of https://github.com/GTNewHorizons/GTplusplus/pull/677. well more precisely quetzl later said they were going to add various tiered versions to GT (not gt++) with a base version for like LuV that has just one solid and one fluid input and versions with more slots for later tiers. Guess we should ask them what the status is. Yours should then be tiered after all these.

**Comment by Alastors** [comment]
> for a recipe this needs to be much more late game. it trivializes a lot.
> 
> 
> 
> Its an even more powerful version of https://github.com/GTNewHorizons/GTplusplus/pull/677. well more precisely quetzl later said they were going to add various tiered versions to GT (not gt++) with a base version for like LuV that has just one solid and one fluid input and versions with more slots for later tiers. Guess we should ask them what the status is. Yours should then be tiered after all these.

Nah I'd like to see at least a form of this in EV, even if it can only hold one recipe or something

**Comment by Alastors** [comment]
We can give this more negatives early game to make it equal, but there are just too many boons to having at least some form of this earlier in the game

**Comment by Dream-Master** [comment]
in ev you start with ae why not move it to iv ? 
@harrynull what your recipe ideas here ?

**Comment by harrynull** [comment]
> in ev you start with ae why not move it to iv ? @harrynull what your recipe ideas here ?

I'm very open to ideas / suggestions for recipe.

My current idea is `1 Stocking Input Bus + 1 Input Hatch (EV) + 3 Pattern Capacity Card + 1 Dual Interface` which tiers this into EV, but I can see how it could be too overpowered.

Alternatively this can be tiered but I can't think of any appropriate limitation other than a smaller # of max pattern.

**Comment by Alastors** [comment]
> in ev you start with ae why not move it to iv ?
> @harrynull what your recipe ideas here ?

I mean honestly that’s fine by me, I’d just like to see at least a version of this around the time you’re able to get AE

**Comment by BlueWeabo** [comment]
What could be done is tier it from EV+
EV one could be just min patterns, like 9 of them.
Then from IV you get 9 more. until you get to the max of 36 at ZPM, which IMO is fine. (could also require a higher tier input hatch)
The way to tier it would probably be with an assembler craft with minimum voltage, which is how the input hatches are tiered.

**Comment by Runakai1** [comment]
This is super cool. I do not agree with the others that this should be EV. I'd rather see it in UV/UHV when machine spam really Happens

**Comment by Alaking** [comment]
Maybe it could be tiered early but paired with a really high energy cost to run/idle cost, so using it becomes a major inconvenience until the neutronium energy csll

**Comment by chochem** [comment]
> 
> I'm very open to ideas / suggestions for recipe.
> 
> My current idea is `1 Stocking Input Bus + 1 Input Hatch (EV) + 3 Pattern Capacity Card + 1 Dual Interface` which tiers this into EV, but I can see how it could be too overpowered.
> 
> If you want it to be _really_ late game then you could put infinity chest somewhere which makes it UEV, but it might be too extreme.
> 
> Alternatively this can be tiered but I can't think of any appropriate limitation other than a smaller # of max pattern.

yea earlier versions sounds tricky. number of patterns wouldn't really change much.
Actually in its current form infinity chest sounds kinda appropriate as that is what it replaces with the infinite buffers. But while it does replaces the entire subnet, infinity chest and automation, it also has downsides as you point it. UV/UHV is probably fine.

well there is already a bunch of disagreement on the recipe, we can also continue that some other time as its not included here anyway.

**Comment by Alrightsc** [comment]
Aside from tier disagreements, want to say this is a super cool GT/AE thing that I think needs to be in game

**Comment by Alastors** [comment]
With all of the people I've talked about, there have been 2 main concerns.

1. Tiering this instead of making it all one hatch

2. That infinitely large internal buffer.

I think we should tier it and make it not have an infinitely large internal buffer until much later on, if we did that, we could realistically make this start at like LuV or ZPM

**Comment by Runakai1** [comment]
I agree with alright. 
Personally, having it have infinite storage capabilities per pattern makes it a VERY strong lang reducing machine. Thus, it shouldn't be available below uv/uhv. Basically an entire me chest setup/Inf chest setup. And it can even hold infinite fluids too! 

If it's made tiered, then I don't disagree with making it available at EV and unlocking the infinite capabilities in uv/uhv. Nonetheless, if neither side can agree, I think the opinion of the PR author is the superior opinion and should be considered. Because this DEFINITELY needs to be in game. Even if we can't agree on how it should be tiered.

**Comment by Alrightsc** [comment]
I'd be for splitting the buffer into two parts, one regular interface that takes only non-fluid patterns, which could be tiered in LuV, and upgraded to fluid capability that's tiered some late game so that it can't be abused immediately to trivialize LPF/PA/PAss setups, etc.

**Comment by boubou19** [comment]
For me it's a big nope. you trivialize the automation like crazy. When AE2-FC was introduced it already trivialized the automation, if we go down this trivialization path, then this must be gated at least post asslines. Also, has anyone tested this with a large AE2 system? what are the lag reports? (i don't want guesses, i want measures)

**Comment by Runakai1** [comment]
I thought about it a bit more. This block needs to be UEV+. People need to have made sub nets before they get absolutely trivialized by this. Also it needs to have a blacklist for stuff like EOH. 

Opinions?

**Comment by boubou19** [comment]
i think we can all agree that players must have faced automation challenges before getting access to this kind of stuff. Now at what point do we consider this to be enough?

@Runakai1 if you use AE2 automation for asslines, CALs or chemplants, you already do subnetting. 

My main concern is that given the huge QoL this represents, players will spam this a lot. So we must make sure it doesn't nuke the server stability when used en masse.

**Comment by Runakai1** [comment]
> i think we can all agree that players must have faced automation challenges before getting access to this kind of stuff. Now at what point do we consider this to be enough?
> 
> @Runakai1 if you use AE2 automation for asslines, CALs or chemplants, you already do subnetting.
> 
> My main concern is that given the huge QoL this represents, players will spam this a lot. So we must make sure it doesn't nuke the server stability when used en masse.

I meant the automation challenges that come. Parallelization in huge quantities, CoAl automation, how to make compact designs so your sub netting isn't laggy. 

All that doesn't really begin below UHV and is a critical skill

**Comment by harrynull** [comment]
What about an item-only version available somewhere between EV-IV, and then item/fluid one in UEV.

**Comment by harrynull** [comment]
> Also, has anyone tested this with a large AE2 system? what are the lag reports? (i don't want guesses, i want measures)

I can start testing this in my own server and I welcome other people to try it

**Comment by boubou19** [comment]
> What about an item-only version available somewhere between EV-IV, and then item/fluid one in UEV.

no, it's really this much powerful. Automation is the core of GT:NH, it's not something we can just change like that without having dire consequences on the game experience.

**Comment by Hiagg** [comment]
i do not disagree with uev gating on this with the features listed in the first 3 comments from harry. especially the one where the whole subnet shares the pattern. i support this feature though :). also if this bus works on assline or any other ordered input multis, it definitely makes sense to put it beyond when youve had them automated for a bit

**Comment by harrynull** [comment]
> > What about an item-only version available somewhere between EV-IV, and then item/fluid one in UEV.
> 
> no, it's really this much powerful. Automation is the core of GT:NH, it's not something we can just change like that without having dire consequences on the game experience.

The item only version isn't very different from interface - super bus combination. I could even change it to use a super bus (max) or something if makes more sense. It doesn't trivialize anything imo.

The dual version is a very important QoL in late tiers and I think it makes sense to make players use it in UEV after they suffer through setting up tens or hundreds of subnets.

**Comment by BlueWeabo** [comment]
I would have to agree with boubou. At first when i mentioned it beign for lower tiers I didn't understand that each pattern was its own inventory. So i would say that UEV would be fine for it. SHould be a reward for those who have played for long.

**Comment by Runakai1** [comment]
The item only version for lower tiers (DEFINITELY NOT EV) would be fine as it would replace only a me chest with storage bus setup. But that needs to be argued as well because it is still trivial with an item only block.

The fluid + item version needs to be UEV+ (and possibly blacklisted for EOH and anything else that should require a complex automation setup in endgame)

**Comment by harrynull** [comment]
Yea I don't disagree. What about gating the item-only version after assline?

Additionally, the original idea where pattern inventory is shared within ME network ends up being too hard to implement, and it has a lot of things to consider: what if the bus is disconnected, or if the network has changed (divided / combined)

Therefore I think an easier implementation is to become centralized, and add an additional hatch called Crafting Input Slave or something, which can be bind to a buffer to share the same inventories. We can tier this slave thing separately too.

**Comment by Runakai1** [comment]
Item version after assline, item + fluid version in uev+ imo.

**Comment by Alrightsc** [comment]
Agreed - item-only patterns version post-AL, don't have a specific preference to fluid gating (UEV+ is fine)

**Comment by JBadges** [comment]
If this block has an internal inventory only seen to its specific machine will this not increase sim times in end game due to patterns needing to be smaller to parallelize this across multiple machines effectively? in that case this would serve more use in the mid-late game. On the other hand if this is across all machines with the same pattern this should be gated late, post spam-able infinity chests as this would be the main automation used to craft every on-demand component as it trivializes parallels with fluids.

**Comment by Eranicus** [comment]
I would say UIV with transcended and maybe with + spacetime for the main hatch. This is definitely should not be low tier item. 

As for EoH blacklist , here I would disagree. EoH automation in my opinion is just creating unnesesary lag considering the EoH amount needed is 30+

**Comment by Alastors** [comment]
> I would say UIV with transcended and maybe with + spacetime for the main hatch. This is definitely should not be low tier item. 
> 
> 
> 
> As for EoH blacklist , here I would disagree. EoH automation in my opinion is just creating unnesesary lag considering the EoH amount needed is 30+

Nah lad, let's go all the way, you can't make this until after stargate

**Comment by repo-alt** [comment]
> > What about an item-only version available somewhere between EV-IV, and then item/fluid one in UEV.
> 
> no, it's really this much powerful. Automation is the core of GT:NH, it's not something we can just change like that without having dire consequences on the game experience.

Tbh, automation-wise it is not very different from an interface on an input bus, if we are speaking about item-only version. It just checks recipes faster and allows very big stacks in the patterns.

**Comment by repo-alt** [comment]
btw, it would be nice to have ICraftingProvider.getCrafterIcon implemented to better see those buses in the network tool display

**Comment by harrynull** [comment]
image

Introducing Crafting Input Slave! You can link it to a Crafting Input Buffer with a data stick, and it shares the same internal inventory of the connected buffer, making parallel much easier as you don't have to rely on AE to split the order. This is essentially the same like the subnet + wireless design before, but it handles fluids too which I think would be impossible/very difficiult to do without this. It also does not need to be connected to ME network, does not have UI and does not hold any patterns.

This can be tiered the same or differently than the buffer.

image
@repo-alt done! it shows up as the GT machine in the crafting tree now.
I still need to figure out custom name / showing up in interface terminal part.

**Comment by repo-alt** [comment]
Intreface terminal support need to be added in the terminal itself I'm afraid, may be done later.

**Comment by harrynull** [comment]
Also added `Crafting Input Bus`, which is an item-only version of Crafting Input Buffer. Other than it does not accept any buffer with fluid inputs, it is the same as the buffer one. Lmk if there is a better name for it, if they sound too similar.

**Review by repo-alt — APPROVED** [review]
It's a bit hardto notice possible dupes fromjust the code, so that only testing will show, but in general the logic looks ok.

**Comment by Dream-Master** [comment]
@harrynull now we need to find a recipe right ?

**Comment by harrynull** [comment]
> @harrynull now we need to find a recipe right ?

Yea we need to find a recipe that everyone can agree on.

**Comment by Dream-Master** [comment]
ok so we add a new pr for this i would say

**Comment by harrynull** [comment]
Yea recipes are usually added in CoreMod anyways I think, so this one can be merged without a recipe decided if everything else looks good.

**Comment by Glease** [comment]
there were a few addons that does not use addInputBusToMachineList and override addToMachineList, or simply not use any of these two (mainly gt++ and maybe also tectech). they should all have been migrated over the last time we did survivalAutoPlace but I'm not sure if there is still anything left behind.
what this means if any multi does what I said above, it will effectively not be able to pick this up correctly and instead identify it as a normal input bus. they are probably not using GPL anyway so we will need to migrate it over regardless of this PR. we should probably watch for such cases while migrating.

**Comment by harrynull** [comment]
@Glease I know the PrAss is an offender to this and I could make a PR after this one is merged. But I guess will need to find more as we go. Other than that I think most GT++ and bartworks multi work

Also pretty sure this won't work with EOH anyways because that doesn't use GPL.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 1]
```diff
@@ -694,29 +697,45 @@ public CheckRecipeResult checkProcessing() {
...
+        // check crafting input hatches first
+        for (IDualInputHatch dualInputHatch : mDualInputHatches) {
+            for (var it = dualInputHatch.inventories(); it.hasNext();) {
+                IDualInputInventory slot = it.next();
+                processingLogic.setInputItems(slot.getItemInputs());
+                processingLogic.setInputFluids(slot.getFluidInputs());
+                result = processingLogic.process();
+                if (result.wasSuccessful()) break;
+            }
+            if (result.wasSuccessful()) break;
```
Comment (minecraft7771) [inline src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:709]: Isn't this line unnecessary.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 2]
```diff
@@ -694,29 +697,45 @@ public CheckRecipeResult checkProcessing() {
...
+        }
+
+        processingLogic.setInputFluids(getStoredFluids());
```
Comment (minecraft7771) [inline src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:712]: Doesn't this interfere with item only version? People that passive supply a fluid wouldn't be able to use it right?

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 3]
```diff
@@ -81,9 +78,7 @@
 import gregtech.common.GT_Pollution;
 import gregtech.common.gui.modularui.widget.CheckRecipeResultSyncer;
 import gregtech.common.items.GT_MetaGenerated_Tool_01;
-import gregtech.common.tileentities.machines.GT_MetaTileEntity_Hatch_InputBus_ME;
-import gregtech.common.tileentities.machines.GT_MetaTileEntity_Hatch_OutputBus_ME;
-import gregtech.common.tileentities.machines.GT_MetaTileEntity_Hatch_Output_ME;
+import gregtech.common.tileentities.machines.*;
```
Comment (minecraft7771) [inline src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:81]: Afaik currently we don't want wildcard imports, so on all files you need to revert back to explicit imports.

**Review by minecraft7771 — COMMENTED** [review]
This review is only concerning GPL interaction. I only commented MultiBlockBase but the review holds obviously true for Extended too.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java (hunk 1 above)
Comment (harrynull) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:709]: won't it continue processing other dual input hatches without? so multiple recipes are processed in the same tick? There can be more than one crafting input hatch in one multi.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java (hunk 2 above)
Comment (harrynull) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:712]: Yea you are right. I don't know if I should call it a Feature™ or should somehow makes the item version also takes stored fluid.

The argument is that that'd slow down recipe checks and prevents us from doing further optimizations, e.g. lock recipe for each pattern slot.
However, it is also more intuitive and might be necessary if you want it to use in something like cutting machine with passived lubricant for example.

(This also applies for the dual version where the player deliberately wants to passive some fluids for reasons)

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java (hunk 1 above)
Comment (minecraft7771) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:709]: Between Line 707 and Line 709 I see now way how the result can change or what did I miss?

Comment (minecraft7771) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:709]: O nevermind I didn't see two loops that you want to leave. We good.

**Comment by TimberDimber** [comment]
It looks like the Crafting Input Buffer is mistaking SLOT_MANUAL and SLOT_CIRCUIT for a crafting item and trying to start the controller.  
The first screenshot is the controller in its normal state.
Second screenshot - spikes that occur in it with a periodicity of 10 seconds, but there is no circuit/shape in the Crafting Input Buffer.
image
image
Third screenshot - the same controller, spikes with the same periodicity, but there is a circuit in the Crafting Input Buffer.
image
These splices are getting bigger and bigger with time. After 3 hours of waiting, they have reached the next timing:
image

**Pull request merged**